source/gui/messageDialog.py

```diff
@@ -0,0 +1,722 @@
+		This will be None when no sound should be played.
+		"""
+		match self:
+			case self.ERROR:
+				return winsound.MB_ICONHAND
+			case self.WARNING:
+				return winsound.MB_ICONASTERISK
+			case _:
+				return None
+
+
+class Button(NamedTuple):
+	"""A button to add to a message dialog."""
+
+	id: ReturnCode
+	"""The ID to use for this button.
+
+	This will be returned after showing the dialog modally.
+	It is also used to modify the button later.
+	"""
+
+	label: str
+	"""The label to display on the button."""
+
+	callback: Callback_T | None = None
+	"""The callback to call when the button is clicked."""
+
+	default_focus: bool = False
+	"""Whether this button should explicitly be the default focused button.
+
+	:note: This only overrides the default focus.
+	If no buttons have this property, the first button will be the default focus.
+	"""
+
+	default_action: bool = False
+	"""Whether this button is the default action.
+
+	The default action is called when the user presses escape, the title bar close button, or the system menu close item.
+	It is also called when programatically closing the dialog, such as when shutting down NVDA.
+
+	:note: This only sets whether to override the default action.
+		EscapeCode.DEFAULT may still result in this button being the default action.
+	"""
+
+	closes_dialog: bool = True
+	"""Whether this button should close the dialog when clicked.
+
+	:note: Buttons with default_action=True and closes_dialog=False are not supported.
+		See the documentation of c{MessageDialog} for information on how these buttons are handled.
+	"""
+
+
+class DefaultButton(Button, Enum):
+	"""Default buttons for message dialogs."""
+
+	# Translators: An ok button on a message dialog.
+	OK = Button(id=ReturnCode.OK, label=_("OK"))
+	# Translators: A yes button on a message dialog.
+	YES = Button(id=ReturnCode.YES, label=_("&Yes"))
+	# Translators: A no button on a message dialog.
+	NO = Button(id=ReturnCode.NO, label=_("&No"))
+	# Translators: A cancel button on a message dialog.
+	CANCEL = Button(id=ReturnCode.CANCEL, label=_("Cancel"))
+	# Translators: A save button on a message dialog.
+	SAVE = Button(id=ReturnCode.SAVE, label=_("&Save"))
+	# Translators: An apply button on a message dialog.
+	APPLY = Button(id=ReturnCode.APPLY, label=_("&Apply"), closes_dialog=False)
+	# Translators: A close button on a message dialog.
+	CLOSE = Button(id=ReturnCode.CLOSE, label=_("Close"))
+	# Translators: A help button on a message dialog.
+	HELP = Button(id=ReturnCode.HELP, label=_("Help"), closes_dialog=False)
+
+
+class DefaultButtonSet(tuple[DefaultButton], Enum):
+	OK_CANCEL = (
+		DefaultButton.OK,
+		DefaultButton.CANCEL,
+	)
+	YES_NO = (
+		DefaultButton.YES,
+		DefaultButton.NO,
+	)
+	YES_NO_CANCEL = (
+		DefaultButton.YES,
+		DefaultButton.NO,
+		DefaultButton.CANCEL,
+	)
+	SAVE_NO_CANCEL = (
+		DefaultButton.SAVE,
+		# Translators: A don't save button on a message dialog.
+		DefaultButton.NO.value._replace(label=_("Do&n't save")),
```

Comment:
I think it might be better to make this a different button

Reply:
This was chosen because of the value used in [`wx.StdDialogButtonSizer`](https://docs.wxpython.org/wx.StdDialogButtonSizer.html), the documentation for which says:

> If you add a button with `wx.ID_SAVE`, on macOS the button will be renamed to “Save” and the `wx.ID_NO` button will be renamed to “Don’t Save” in accordance with the macOS Human Interface Guidelines.

Which implies to me that `wx.ID_SAVE` and `wx.ID_NO` is the most appropriate pairing of IDs in this situation.